python-scripts/eeprom-util.py

```diff
@@ -0,0 +1,198 @@
+    self.confounder = bytes(eepFile.read(8))      # encrypted confounder
+    self.hddKey = bytes(eepFile.read(16))         # encrypted HDD key
+    self.gameReg = bytes(eepFile.read(4))         # encrypted game region code
+
+    # identity variables
+    self.identChk = bytes(eepFile.read(4))        # identity checksum
+    self.serial = bytes(eepFile.read(12))         # serial number
+    self.mac = bytes(eepFile.read(6))             # MAC address
+    self.unknown0 = bytes(eepFile.read(2))        # TBD
+    self.onlineKey = bytes(eepFile.read(16))      # online key
+    self.videoStd = bytes(eepFile.read(4))        # NTSC / PAL
+    self.unknown1 = bytes(eepFile.read(4))        # TBD
+
+    # settings variables
+    self.settChk = bytes(eepFile.read(4))         # settings checksum
+    self.timeZoneBias = bytes(eepFile.read(4))    # time zone bias
+    self.timeZoneStdName = bytes(eepFile.read(4)) # standard timezone
+    self.timeZoneDltName = bytes(eepFile.read(4)) # daylight savings timezone
+    self.unknown2 = bytes(eepFile.read(8))        # TBD
+    self.timeZoneStdDate = bytes(eepFile.read(4)) # standard date / time
+    self.timeZoneDltDate = bytes(eepFile.read(4)) # daylight savings date / time
+    self.unknown3 = bytes(eepFile.read(8))        # TBD
+    self.timeZoneStdBias = bytes(eepFile.read(4)) # standard bias
+    self.timeZoneDltBias = bytes(eepFile.read(4)) # daylight savings bias
+    self.languageId = bytes(eepFile.read(4))      # language identifier
+    self.videoFlags = bytes(eepFile.read(4))      # video settings
+    self.audioFlags = bytes(eepFile.read(4))      # audio settings
+    self.gameParCont = bytes(eepFile.read(4))     # game parental contol settings
+    self.parContPass = bytes(eepFile.read(4))     # parental control password
+    self.dvdParCont = bytes(eepFile.read(4))      # DVD parental control settings
+    self.ip = bytes(eepFile.read(4))              # IP address
+    self.dns = bytes(eepFile.read(4))             # DNS server
+    self.gateway = bytes(eepFile.read(4))         # default gateway
+    self.subnet = bytes(eepFile.read(4))          # subnet
+    self.unknown4 = bytes(eepFile.read(4))        # TBD
+    self.dvdReg = bytes(eepFile.read(4))          # DVD region code
+    self.unknown5 = bytes(eepFile.read(64))       # TBD
+
+# Because of the confounder we have to split the RC4 decryption / encryption proccess
+# into two functions (eepKeyPrepCrypt / eepKeyCrypt). We also need a way to share state
+# between the the two functions (EepKeyObj). In a nutshell we fill and EepKeyObj method
+# object with the initial state by calling eepKeyPrepCrypt. Then we call eepKeyCrypt
+# on the confounder data. eepKeyCrypt will push state data back into the EepKeyObj method 
+# object (i / j) which is necessary when we call eepKeyCrypt on the next chunk of data
+# from the Eeprom method object (eeprom.hddKey + eeprom.gameReg). This is also why we
+# don't use any standard RC4 crypto libaries here.
```

Comment:
https://en.wikipedia.org/wiki/RC4 - This was written based on that

You have two RC4 stages, the KSA and PRGA. The KSA is initializing a permutation of the key, The PRGA actually does the encryption / decryption. Consider the following based on a standard RC4 algorithm.

Normally It goes like this.

```
def rc4 (key, data):
  keyPerm = KSA(key) # generates a permutation using the key
  decData = PGRA(keyPerm, data) # decrypts data using the key permutation 
  return decData
```

So if we try to encrypt an eeprom with this function this is what happens.

```
confounder = rc4 (eepromKey, encryptedConfouder)

def rc4 (key, data):
  keyPerm = KSA(key)
  decData = PGRA(keyPerm, data)
  return decData
```

This part works, we get our confounder, but doing it again will not work for the hdd key / game reg

```
hddKeyGameReg = rc4 (eepromKey, encryptedBytes)

def rc4 (key, data):
  #after the last function ended we lost the key permutation, no issues we'll make it again
  #except during the PGRA stage we stored I and J from decrypting the confounder in the
  # permutation. We need I and J from the confounder PGRA stage to decrypt further .
  # So keyPerm needs to be a seperate entitiy. Also KSA need to only be called once or
  # it wipe data from the previous PGRA stage as well.
  keyPerm = KSA(key)

  # By default this function set I and J to zero at start, we have to change that as it needs to 
  # pull those values from keyPerm. If you don't do this it wont work.
  decData = PGRA(keyPerm, data)
  return decData
```

So keyPerm, KSA, and PGRA all must be changed from their default behaviour in order to account for the confounder.

keyPerm data has to persist between function calls, so it must sit outside of the functions.
KSA must only be once, when we first decrypt the confounder
PGRA must pull starting values from keyPerm and export the finishing values back to keyPerm

It's hard to follow i know, i beat my head against a wall four hours figuring it out, but that's why the standard RC4 functions don't work, because they don't consider that the confounder data has to stick around through the rest of the process.